### PR TITLE
Add PyTorch Lazy Initialization

### PR DESCRIPTION
**Pull request by astonzhang**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_builders-guide/lazy-init.md [hunk 1]
```diff
@@ -169,6 +168,22 @@ but the framework initializes sequentially.
 Once all parameter shapes are known,
 the framework can finally initialize the parameters.
 
+:begin_tab:`pytorch`
+The following method
+passes in an input
+to infer all parameter shapes
+before initializing parameters.
```
Comment (AnirudhDagar) [inline chapter_builders-guide/lazy-init.md:175]:
```suggestion
The following method
passes in a dummy input
through the network
for a dry-run
to infer all parameter shapes
and subsequently initializes the parameters.
```

**Review by AnirudhDagar — APPROVED** [review]
This looks much better than hard coding. I've made a minor comment for explaining the dry-run step explicitly.

Btw `chapter_convolutional-neural-networks/lenet.md` shows tab execution numbers. `d2lbook activate all chapter_convolutional-neural-networks/lenet.md` should be run.

File: chapter_builders-guide/lazy-init.md (hunk 1 above)
Comment (astonzhang) [reply chapter_builders-guide/lazy-init.md:175]: This reads better. Thanks

File: chapter_builders-guide/lazy-init.md [hunk 2]
```diff
@@ -169,6 +168,24 @@ but the framework initializes sequentially.
...
+for a dry run
+to infer all parameter shapes
+and subsequently initializes the parameters.
+It will be used later when default random initializations are not desired.
+:end_tab:
+
+```{.python .input}
+%%tab pytorch
+@d2l.add_to_class(d2l.Module)  #@save
+def apply_init(self, init, X):
```
[the remark this answers is not in the document]
Comment (cheungdaven) [reply chapter_builders-guide/lazy-init.md:184]: X should be a list or a tuple to accommodate cases where multiple inputs are required.

[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_builders-guide/lazy-init.md:184]: In this case let's call `X` as `inputs` (use plural to indicate a container for better readability)

**Review by cheungdaven — APPROVED** [review]

**Pull request merged**